src/micro.c

```diff
@@ -0,0 +1,680 @@
+    {
+        (*m->prev_on_connection_closed)(nc, m->prev_on_connection_disconnected_closure);
+    }
+}
+
+static void
+on_error(natsConnection *nc, natsSubscription *sub, natsStatus s, void *closure)
+{
+    microService *m = (microService *)closure;
+    microEndpoint *ep = NULL;
+    microError *err = NULL;
+    bool our_subject = false;
+
+    if ((m == NULL) || (sub == NULL))
+        return;
+
+    micro_lock_service(m);
+    for (ep = m->first_ep; (!our_subject) && (ep != NULL); ep = ep->next)
+    {
+        if (micro_match_endpoint_subject(ep->subject, sub->subject))
+        {
+            break;
+        }
+    }
+    micro_unlock_service(m);
+
+    if (m->cfg->ErrHandler != NULL)
+    {
+        (*m->cfg->ErrHandler)(m, ep, s);
+    }
+
+    if (ep != NULL)
+    {
+        err = microError_Wrapf(micro_ErrorFromStatus(s), "NATS error on endpoint %s", ep->name);
+        micro_update_last_error(ep, err);
+        microError_Destroy(err);
+    }
+
+    // <>/<> TODO: Should we stop the service? The Go client does.
+    microService_Stop(m);
+
+    if (m->prev_on_error != NULL)
+    {
+        (*m->prev_on_error)(nc, sub, s, m->prev_on_error_closure);
+    }
+}
+
+static microError *
+wrap_connection_event_callbacks(microService *m)
+{
+    natsStatus s = NATS_OK;
+
+    if (m == NULL)
+        return micro_ErrorInvalidArg;
+
+    IFOK(s, natsConn_getClosedCallback(&m->prev_on_connection_closed, &m->prev_on_connection_closed_closure, m->nc));
+    IFOK(s, natsConn_setClosedCallback(m->nc, on_connection_closed, m));
+
+    IFOK(s, natsConn_getDisconnectedCallback(&m->prev_on_connection_disconnected, &m->prev_on_connection_disconnected_closure, m->nc));
+    IFOK(s, natsConn_setDisconnectedCallback(m->nc, on_connection_disconnected, m));
+
+    IFOK(s, natsConn_getErrorCallback(&m->prev_on_error, &m->prev_on_error_closure, m->nc));
+    IFOK(s, natsConn_setErrorCallback(m->nc, on_error, m));
+
+    return microError_Wrapf(micro_ErrorFromStatus(s), "failed to wrap connection event callbacks");
+}
+
+static microError *
+unwrap_connection_event_callbacks(microService *m)
+{
+    natsStatus s = NATS_OK;
+
+    if (m == NULL)
+        return micro_ErrorInvalidArg;
+
+    IFOK(s, natsConn_setClosedCallback(m->nc, m->prev_on_connection_closed, m->prev_on_connection_closed_closure));
+    IFOK(s, natsConn_setDisconnectedCallback(m->nc, m->prev_on_connection_disconnected, m->prev_on_connection_disconnected_closure));
+    IFOK(s, natsConn_setErrorCallback(m->nc, m->prev_on_error, m->prev_on_error_closure));
+
+    return microError_Wrapf(micro_ErrorFromStatus(s), "failed to unwrap connection event callbacks");
+}
+
+microError *
+microService_AddGroup(microGroup **new_group, microService *m, const char *prefix)
+{
+    if ((m == NULL) || (new_group == NULL) || (prefix == NULL))
+        return micro_ErrorInvalidArg;
+
+    *new_group = NATS_CALLOC(1, sizeof(microGroup) +
+                                    strlen(prefix) + 1); // "prefix.""
```

Comment:
Is `prefix` already `prefix.` or is the `+1` for the dot? If so, should you add 1 more for the `\0`?